core/src/components/input-password-toggle/input-password-toggle.tsx

```diff
@@ -0,0 +1,140 @@
+
+/**
+ * @virtualProp {"ios" | "md"} mode - The mode determines which platform styles to use.
+ */
+@Component({
+  tag: 'ion-input-password-toggle',
+  styleUrls: {
+    ios: 'input-password-toggle.ios.scss',
+    md: 'input-password-toggle.md.scss',
+  },
+  shadow: true,
+})
+export class InputPasswordToggle implements ComponentInterface {
+  private inputElRef!: HTMLIonInputElement | null;
+
+  @Element() el!: HTMLIonInputElement;
+
+  /**
+   * The color to use from your application's color palette.
+   * Default options are: `"primary"`, `"secondary"`, `"tertiary"`, `"success"`, `"warning"`, `"danger"`, `"light"`, `"medium"`, and `"dark"`.
+   * For more information on colors, see [theming](/docs/theming/basics).
+   */
+  @Prop({ reflect: true }) color?: Color;
+
+  /**
+   * The icon that can be used to represent showing a password. If not set, the "eye" Ionicon will be used.
+   */
+  @Prop() showIcon?: string;
+
+  /**
+   * The icon that can be used to represent hiding a password. If not set, the "eyeOff" Ionicon will be used.
+   */
+  @Prop() hideIcon?: string;
+
+  connectedCallback() {
+    const { el } = this;
+
+    const inputElRef = (this.inputElRef = el.closest('ion-input'));
+
+    if (!inputElRef) {
+      printIonWarning(
+        'No ancestor ion-input found for ion-input-password-toggle. This component must be slotted inside of an ion-input.',
+        el
+      );
+
+      return;
+    }
+
+    inputElRef.addEventListener('ionTypeChange', this.onTypeChange);
+  }
+
+  disconnectedCallback() {
+    this.inputElRef?.removeEventListener('ionTypeChange', this.onTypeChange);
+    this.inputElRef = null;
+  }
+
+  private togglePasswordVisibility = () => {
+    const { inputElRef } = this;
+
+    if (!inputElRef) {
+      return;
+    }
+
+    inputElRef.type = inputElRef.type === 'text' ? 'password' : 'text';
+  };
+
+  /**
+   * Whenever the input type changes we need to re-run validation to ensure the password
+   * toggle is being used with the correct input type. If the application changes the type
+   * outside of this component we also need to re-render so the correct icon is shown.
+   */
+  private onTypeChange = () => {
+    const { el, inputElRef } = this;
+
+    if (!inputElRef) return;
+
+    if (inputElRef.type !== 'text' && inputElRef.type !== 'password') {
+      printIonWarning(
+        `ion-input-password-toggle only supports inputs that accept plain text. Input of type "${inputElRef.type}" is not compatible.`,
+        el
+      );
+
+      return;
+    }
+
+    forceUpdate(this);
+  };
+
+  render() {
+    const { color, inputElRef } = this;
+
+    const mode = getIonMode(this);
+
+    const showPasswordIcon = this.showIcon ?? eye;
+    const hidePasswordIcon = this.hideIcon ?? eyeOff;
+
+    const isPasswordVisible = inputElRef?.type === 'text';
+
+    return (
+      <Host
+        class={createColorClasses(color, {
+          [mode]: true,
+        })}
+      >
+        <ion-button
```

Comment:
I noticed that the button doesn't show a ripple effect in MD; was that intentional?

Reply:
Not intentional, but not 100% sure why that would be happening. I'll look into it.

Reply:
Ok we're going to test this out in https://github.com/ionic-team/ionic-framework/pull/29192

Reply:
Should be fixed in this branch now!